### PR TITLE
FFMQ: Efficiency Improvement and Use New Options Methods

### PR DESCRIPTION
**Pull request by Alchav**
## What is this fixing or adding?
- Converts rooms.yaml and entrances.yaml to python data so no yaml loading and converting takes place during runtime while the apworld is initialized. This saves a good few seconds of loading time.
- Updates FFMQ to use new options methods
- Removes some `print` statements that were left in by mistake.
- Fixes a bug where one option was being checked twice and two others not checked at all, for determining if a map shuffle API call was necessary.
- Entrance hints will now show locations at Mac's Ship (why did I go out of my way to exclude it? It is a mystery)

## How was this tested?
Generating several times with various options.

### CODE REVIEW

File: worlds/ffmq/Client.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ async def game_watcher(self, ctx):
         received = await snes_read(ctx, RECEIVED_DATA[0], RECEIVED_DATA[1])
         data = await snes_read(ctx, READ_DATA_START, READ_DATA_END - READ_DATA_START)
         check_2 = await snes_read(ctx, 0xF53749, 1)
-        if check_1 in (b'\x00', b'\x55') or check_2 in (b'\x00', b'\x55'):
+        if check_1 != b'01' or check_2 != b'01':
```
Comment (Zunawe) [inline worlds/ffmq/Client.py:74]: I assume this was snuck in as a quick simplification? Doesn't change any behaviors or anything?

**Review by Zunawe — COMMENTED** [review]
- `Output.py` line 102 uses `self.multiworld.sky_coin_mode[self.player]`
- `stage_generate_early` could (should?) access options through the `world` variable

Otherwise, does what it says on the tin.

World load on my machine goes from
- ~0.75s to ~0.075s on Python 3.8
- ~0.55s to ~0.08s on Python 3.11

Did not run any code except to import the world.

File: worlds/ffmq/Client.py (hunk 1 above)
Comment (Alchav) [reply worlds/ffmq/Client.py:74]: It makes it less likely that hard resetting in bizhawk will erroneously send checks or game completion.

**Comment by Alchav** [comment]
> * `Output.py` line 102 uses `self.multiworld.sky_coin_mode[self.player]`
> 
>     * `stage_generate_early` could (should?) access options through the `world` variable
> 
>     * `Regions.py` line 70 uses `self.multiworld.brown_boxes[self.player]`
> 
>     * `Regions.py` line 213 uses `multiworld.enemies_density[player]`
> 
> 
> Otherwise, does what it says on the tin.
> 
> World load on my machine goes from
> 
>     * ~0.75s to ~0.075s on Python 3.8
> 
>     * ~0.55s to ~0.08s on Python 3.11

Fixed

**Review by Zunawe — APPROVED** [review]
Looks like all options access was converted.

File: worlds/ffmq/__init__.py [hunk 2]
```diff
@@ -204,16 +205,14 @@ def extend_hint_information(self, hint_data):
                             for location in exit_check.connected_region.locations:
                                 if location.address:
                                     hint = []
-                                    if self.multiworld.map_shuffle[self.player] != "dungeons":
+                                    if self.options.map_shuffle != "dungeons":
                                         hint.append((subregion.split("Subregion ")[-1] + (" Region" if subregion not
                                                     in single_location_regions else "")))
-                                    if self.multiworld.map_shuffle[self.player] != "overworld" and subregion not in \
-                                            ("Subregion Mac's Ship", "Subregion Doom Castle"):
+                                    if self.options.map_shuffle != "overworld":
```
Comment (Exempt-Medic) [inline worlds/ffmq/__init__.py:211]: Just checking that this subregion removal is intended here.

**Review by Exempt-Medic — APPROVED** [review]
I had one comment on a removed bit of code. Otherwise the changes all LGTM. Merged into main and generated ten multiworlds with 100 random yamls. Everything generated fine and did not hit any deprecated options warning. Searched through the code as well and did not find any instances of that or per_slot_randoms.

File: worlds/ffmq/__init__.py (hunk 2 above)
Comment (Alchav) [reply worlds/ffmq/__init__.py:211]: > Just checking that the removal of the subregion check here is intended.

From the description of this PR:

> Entrance hints will now show locations at Mac's Ship (why did I go out of my way to exclude it? It is a mystery)

Comment (Exempt-Medic) [reply worlds/ffmq/__init__.py:211]: And I wasn't sure if Doom Castle was relevant to that. But by your answer, I'll just assume it was

File: worlds/ffmq/__init__.py [hunk 3]
```diff
@@ -69,20 +70,20 @@ def __init__(self, world, player: int):
...
-        if self.multiworld.bosses_scaling_lower[self.player].value > \
-                self.multiworld.bosses_scaling_upper[self.player].value:
-            (self.multiworld.bosses_scaling_lower[self.player].value,
-             self.multiworld.bosses_scaling_upper[self.player].value) =\
-                (self.multiworld.bosses_scaling_upper[self.player].value,
-                 self.multiworld.bosses_scaling_lower[self.player].value)
+        if self.options.sky_coin_mode == "shattered_sky_coin":
+            self.options.brown_boxes.value = 1
+        if self.options.enemies_scaling_lower.value > \
+                self.options.enemies_scaling_upper.value:
```
Comment (NewSoupVi) [inline worlds/ffmq/__init__.py:76]: This is short enough to be on one line now, no? :)

Same with a lot of the following lines

**Review by NewSoupVi — CHANGES_REQUESTED** [review]
Ready to merge, just noticed one small thing

File: worlds/ffmq/Regions.py [hunk 4]
```diff
@@ -1,11 +1,9 @@
 from BaseClasses import Region, MultiWorld, Entrance, Location, LocationProgressType, ItemClassification
 from worlds.generic.Rules import add_rule
+from .data.rooms import rooms, entrances
 from .Items import item_groups, yaml_item
-import pkgutil
-import yaml
 
-rooms = yaml.load(pkgutil.get_data(__name__, "data/rooms.yaml"), yaml.Loader)
```
Comment (black-sliver) [inline worlds/ffmq/Regions.py:7]: I think we should be more thorough in reviews (and hopefully some peer reviewers see this): for yaml this should've been `Utils.parse_yaml`, which is both safer and 7x faster than this. There are very few reasons to use anything else.

File: worlds/ffmq/__init__.py (hunk 3 above)
Comment (black-sliver) [reply worlds/ffmq/__init__.py:76]: +1
It's really hard to read. also the (..., ...) = (..., ...) assignment below is even worse with the breaks. If it needs breaks, there is no reason to do tuple assignments like this.

**Review by black-sliver — COMMENTED** [review]
Can you change `settings_template` in `Output.py` to use `Utils.parse_yaml`? (see below)

**Comment by Alchav** [comment]
> Can you change `settings_template` in `Output.py` to use `Utils.parse_yaml`? (see below)

Done

File: worlds/ffmq/__init__.py (hunk 3 above)
Comment (Alchav) [reply worlds/ffmq/__init__.py]: Changed.

**Review by NewSoupVi — APPROVED** [review]

**Comment by TreZc0** [comment]
This should be reverted, as it currently breaks SNI connection on FFMQ.
The reason is the one line change in client.py

**Comment by Exempt-Medic** [comment]
@Alchav can just make another PR, no?

**Pull request merged**